src/main/java/net/minecraftforge/common/util/SortedRegistry.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.util;
+
+import com.google.common.collect.*;
+import com.google.common.graph.ElementOrder;
+import com.google.common.graph.GraphBuilder;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraftforge.fml.loading.toposort.TopologicalSort;
+
+import javax.annotation.Nullable;
+import java.util.*;
+
+/**
+ * A {@link ResourceLocation} to T registry where elements may define arbitrary ordering rules.
```

Comment:
I think the javadoc could be reworded somewhat:
```suggestion
 * A registry of elements with {@link ResourceLocation} names with topological sorting.
```